File: src/main/java/com/sisqueslabs/nexora/api/contexts/chat/transport/rest/ChatCompletionsController.java
package com.sisqueslabs.nexora.api.contexts.chat.transport.rest;

import org.axonframework.commandhandling.gateway.CommandGateway;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.sisqueslabs.nexora.api.contexts.chat.application.command.createchatcompletion.CreateChatCompletionCommand;
import com.sisqueslabs.nexora.api.contexts.chat.application.command.createchatcompletion.CreateChatCompletionResult;
import com.sisqueslabs.nexora.api.contexts.chat.transport.rest.dto.ChatCompletionRequestDto;
import com.sisqueslabs.nexora.api.contexts.chat.transport.rest.dto.ChatCompletionResponseDto;

@RestController
@RequestMapping("/v1/chat")
public class ChatCompletionsController {

    private final CommandGateway commandGateway;

    public ChatCompletionsController(CommandGateway commandGateway) {
        this.commandGateway = commandGateway;
    }

    @PostMapping("/completions")
    public ChatCompletionResponseDto createChatCompletion(@RequestBody ChatCompletionRequestDto requestDto) {
        var request = requestDto.toDomain();

        CreateChatCompletionResult result = commandGateway.sendAndWait(new CreateChatCompletionCommand(request));

        return ChatCompletionResponseDto.from(result, request.model());
    }
}

File: src/contexts/chat/transport/rest/controllers/chat-completions.controller.ts
import { CreateChatCompletionCommand } from '@contexts/chat/application/commands/create-chat-completion/create-chat-completion.command';
import { CreateChatCompletionResult } from '@contexts/chat/application/commands/create-chat-completion/create-chat-completion.handler';
import { Body, Controller, HttpCode, HttpStatus, Post } from '@nestjs/common';
import { ApiOperation, ApiResponse, ApiTags } from '@nestjs/swagger';
import { CommandBus } from '@nestjs/cqrs';

import { ChatCompletionRequestDto } from '../dtos/chat-completion-request.dto';
import { ChatCompletionResponseDto } from '../dtos/chat-completion-response.dto';

@ApiTags('chat')
@Controller('v1/chat')
export class ChatCompletionsController {
  constructor(private readonly commandBus: CommandBus) {}

  @Post('completions')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({
    summary: 'Run inference against the (mocked, in v0) platform',
  })
  @ApiResponse({ status: 200, type: ChatCompletionResponseDto })
  async createChatCompletion(
    @Body() requestDto: ChatCompletionRequestDto,
  ): Promise<ChatCompletionResponseDto> {
    const command = new CreateChatCompletionCommand(
      requestDto.toCommandInput(),
    );

    const result = await this.commandBus.execute<
      CreateChatCompletionCommand,
      CreateChatCompletionResult
    >(command);

    return ChatCompletionResponseDto.from(result, command.request.model.value);
  }
}
